keps/sig-autoscaling/2556-circumvent-50-tag-asg-limit/README.md

```diff
@@ -0,0 +1,139 @@
+      "nodegroupArn": "*string*",
+      "nodegroupName": "*string*",
+      "nodeRole": "*string*",
+      "releaseVersion": "*string*",
+      "remoteAccess": { 
+         "ec2SshKey": "*string*",
+         "sourceSecurityGroups": [ "*string*" ]
+      },
+      "resources": { 
+         "autoScalingGroups": [ 
+            { 
+               "name": "*string*"
+            }
+         ],
+         "remoteAccessSecurityGroup": "*string*"
+      },
+      "scalingConfig": { 
+         "desiredSize": *number*,
+         "maxSize": *number*,
+         "minSize": *number*
+      },
+      "status": "*string*",
+      "subnets": [ "*string*" ],
+      "tags": { 
+         "*string*" : "*string*" 
+      },
+      "version": "*string*"
+   }
+}
+```
+
+[Current EKS interface]: https://github.com/aws/aws-sdk-go/blob/82e096143fdfb8f52fbeb4ef78d400eea6381ccd/service/eks/eksiface/interface.go#L96
+[DescribeNodegroup API]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html
+[here]: https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeNodegroup.html#API_DescribeNodegroup_ResponseSyntax
+
+## Implementation Proposal
+
+1. In the autoscaling loop, Cluster Autoscaler finds no nodes, so it will build the nodeTemplate from other information.
+1. Cluster Autoscaler will first get some common allocatable resource values. ([allocatable resources code][])
+1. Cluster Autoscaler will then check the ASG tags for the EKS ManagedNodegroups tags. ([ASG tags code][])
+1. If Cluster Autoscaler finds the EKS ManagedNodegroups tags, it will call methods on the ManagedNodegroupCache to get labels, taints, and some other values.
+   1. If the cache has data for the given ManagedNodegroup, it will return that data.
+   1. If the data’s TTL has expired, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+   1. If there is no data cached for the ManagedNodegroup, the DescribeNodegroup API will be called and the data will be added to the cache and returned.
+1. Then Cluster Autoscaler will pull information from the ASG tags like it already does. If duplicate keys are found, the values from the ASG tags are kept.
+1. Cluster Autoscaler will then continue as usual.
+
+[allocatable resources code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L366
+[ASG tags code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L370
+
+## Testing
+
+- [TC1] Test creating a DescribeNodegroup request
+- [TC2] Test pulling values from cache when data exists for the ManagedNodegroups
+- [TC3] Test pulling values from cache and calling DescribeNodegroup API if no data exists
+- [TC4] Test missing permissions and falling back to just ASG tags
+- [TC5] Test duplicate label keys - selecting the value from ASG tags response
+
+## Risks
```

Comment:
From my understanding if we want the cache TTL to be configurable, we'll likely need to add a new configuration flag to the CA, adding another cloud provider specific flag, I think this should be noted somewhere as a drawback to the design as proposed. (I couldn't see a better place to raise this currently.)

Reply:
Sounds good, I'll add that into the doc. Do you think the flag will cause a lot of push-back?

Reply:
I'm not sure how much pushback it'll get, but I think it'll at least be questioned as there's currently only a small number of cloudprovider specific flags and it's adding some configuration complexity (at least in parsing the potential config options) for functionality only of use to a subset of users. Best to call it out explicitly at this stage.

Reply:
After looking into the possible tps for calling DescribeNodegroup after this change and discussing with teammates, I think we shouldn't make the TTL configurable for now. This can be added later on if it becomes a concern. I'll update the kep with this in mind.